builtins: add underscore variant for each index builtin

otan opened the pull request:
Add underscore variants for each spatial predicate builtin, which does
not utilize the index.

Release note (sql change): Add a _ prefix to index backed geospatial
functions, which avoids using the index whilst applying the same
functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51225)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-:-MBq5JIa8W8Sb4LhUJEZ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBq50wUCiibdbmvhdgz:-MBq50wUCiibdbmvhdh-:b-ngjgt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3402)):*
> ```Go
> 		"st_overlaps",
> 		"st_touches",
> 		"st_within",
> ```

Maybe add a comment somewhere that this should be kept in sync with `opt/invertedidx/geo.go`.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-:-MBqCK8NDHC-oHrANjKY:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @otan and @rytaft)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBq50wUCiibdbmvhdgz:-MBqBWRB1r4x3gaRRb4p:b5ttu24) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3402)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe add a comment somewhere that this should be kept in sync with `opt/invertedidx/geo.go`.
</blockquote></details>

Can we just use that map?
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBqBdNv3jXKjac6WL8Z:-MBqBdNv3jXKjac6WL8_:bdqffu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3413)):*
> ```Go
> 			builtin.overloads[i].Info += "\n\nThis function variant will attempt to utilize any available geospatial index."
> 
> 			ovCopy.Info += "\n\nThis function variant does not attempt to utilize any geospatial index."
> ```

nit: simpler to say "This function variant does not utilize any geospatial index."


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51225)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBq50wUCiibdbmvhdgz:-MBqEmWI-sbc-dHxf7V0:b-jlqmyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3402)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can we just use that map?
</blockquote></details>

Do you mean xform/geo.go, or is this an in progress commit?

I think it'll introduce circular dependency, but I can move the map to `geoindex` and all will be ok.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBq50wUCiibdbmvhdgz:-MBqLxNm9cj0NLfYBq7K:bmffpu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3402)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Do you mean xform/geo.go, or is this an in progress commit?

I think it'll introduce circular dependency, but I can move the map to `geoindex` and all will be ok.
</blockquote></details>

yea, I'm moving it as part of https://github.com/cockroachdb/cockroach/pull/51174. `geoindex` is also a good place for it.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBq50wUCiibdbmvhdgz:-MBqMlqU5A7w02CbqMyq:bibei7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3402)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

yea, I'm moving it as part of https://github.com/cockroachdb/cockroach/pull/51174. `geoindex` is also a good place for it.
</blockquote></details>

I like to move it move it ~ Reel 2 Real, 1994

I shall wait until it crosses over.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBq50wUCiibdbmvhdgz:-MBqO3DdBJHS-uUoPj8A:bc52tdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3402)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I like to move it move it ~ Reel 2 Real, 1994

I shall wait until it crosses over.
</blockquote></details>

:laughing: 


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51225)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @otan)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtSoPY8a4BB8Q6ZUP_:-MBtSoPY8a4BB8Q6ZUPa:b-zly8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3413)):*
> ```Go
> 			builtin.overloads[i].Info += "\n\nThis function variant will attempt to utilize any available geospatial index."
> 
> 			ovCopy.Info += "\n\nThis function variant does not attempt to utilize any geospatial index."
> ```

Did you mean to remove "not attempt to"?
___
*[pkg/geo/geoindex/geoindex.go, line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtSAIA0YBjODuSaqyP:-MBtSAIA0YBjODuSaqyQ:bpoe1ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f711e19befb8ec046e7831d6cb419432aa4843/pkg/geo/geoindex/geoindex.go#L27)):*
> ```Go
> // accelerated. Each function implies a certain type of geospatial relationship,
> // which affects how the index is queried as part of a constrained scan or
> // geospatial lookup join. geoRelationshipMap maps the function name to its
> ```

geoRelationshipMap -> RelationshipMap
___
*[pkg/geo/geoindex/geoindex.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtSDLYBYnphRkLx4vv:-MBtSDLYBYnphRkLx4vw:b-gfviy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f711e19befb8ec046e7831d6cb419432aa4843/pkg/geo/geoindex/geoindex.go#L28)):*
> ```Go
> // which affects how the index is queried as part of a constrained scan or
> // geospatial lookup join. geoRelationshipMap maps the function name to its
> // corresponding relationship (Covers, CoveredBy, or Intersects).
> ```

This list should now include DWithin and DFullyWithin
___
*[pkg/geo/geoindex/geoindex.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtS4DhD-0sNxOk-6MN:-MBtS4DhD-0sNxOk-6MO:bpoe1ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f711e19befb8ec046e7831d6cb419432aa4843/pkg/geo/geoindex/geoindex.go#L62)):*
> ```Go
> // commuted.
> //
> // The relationships in the geoRelationshipMap map above only apply when the
> ```

geoRelationshipMap -> RelationshipMap


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51225)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @rytaft)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtSoPY8a4BB8Q6ZUP_:-MBtWiMkD5LNo2zivYfm:b46via4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ba2a190b0e7f7434232d737b54aed47d8aab44/pkg/sql/sem/builtins/geo_builtins.go#L3413)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Did you mean to remove "not attempt to"?
</blockquote></details>

missing a not!
___
*[pkg/geo/geoindex/geoindex.go, line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtSAIA0YBjODuSaqyP:-MBtWhC6CkkuQk2oplWl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f711e19befb8ec046e7831d6cb419432aa4843/pkg/geo/geoindex/geoindex.go#L27)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

geoRelationshipMap -> RelationshipMap
</blockquote></details>

Done.
___
*[pkg/geo/geoindex/geoindex.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtSDLYBYnphRkLx4vv:-MBtWhVBA1wziz3_RFnj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f711e19befb8ec046e7831d6cb419432aa4843/pkg/geo/geoindex/geoindex.go#L28)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This list should now include DWithin and DFullyWithin
</blockquote></details>

Done.
___
*[pkg/geo/geoindex/geoindex.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtS4DhD-0sNxOk-6MN:-MBtWgq07RIJvO7CJkEF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f711e19befb8ec046e7831d6cb419432aa4843/pkg/geo/geoindex/geoindex.go#L62)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

geoRelationshipMap -> RelationshipMap
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51225)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @otan)

___
*[pkg/geo/geoindex/geoindex.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-MBtSDLYBYnphRkLx4vv:-MBtZz6e9Q9zz1IvwGaa:bjxmx1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f711e19befb8ec046e7831d6cb419432aa4843/pkg/geo/geoindex/geoindex.go#L28)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

[nit] FullyWithin -> DFullyWithin


<!-- Sent from Reviewable.io -->

otan committed:
builtins: add underscore variant for each index builtin

Add underscore variants for each spatial predicate builtin, which does
not utilize the index.

Release note (sql change): Add a _ prefix to index backed geospatial
functions, which avoids using the index whilst applying the same
functionality.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51225#-:-MBuKvUC8v4DQX5DnrTK:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=rytaft
